Conditionally mark getter/setter implementations virtual in generated classes

IISResetMe opened the pull request:
Fix https://github.com/PowerShell/PowerShell/issues/8302

When implementing interfaces, PowerShell incorrectly produces non-virtual get/set methods for interface-defined properties.
This commit adds a lookup method for interface-defined properties and marks get/set methods for properties with matching signatures virtual.

## PR Summary

<!-- summarize your PR between here and the checklist -->

## PR Checklist

- [X] [PR has a meaningful title](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
    - Use the present tense and imperative mood when describing your changes
- [X] [Summarized changes](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] [Change is not breaking](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#making-breaking-changes)
- [X] [Make sure all `.h`, `.cpp`, `.cs`, `.ps1` and `.psm1` files have the correct copyright header](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
- [X] This PR is ready to merge and is not [Work in Progress](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---work-in-progress).
    - If the PR is work in progress, please add the prefix `WIP:` to the beginning of the title and remove the prefix when the PR is ready.
- **User-facing changes**
    - [X] Not Applicable
    - **OR**
    - [ ] User-facing [Documentation needed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#pull-request---submission)
        - [ ] Issue filed - Issue link:
- **Testing - New and feature**
    - [] Not Applicable or can only be tested interactively
    - **OR**
    - [ ] [Make sure you've added a new test if existing tests do not effectively test the code changed](https://github.com/PowerShell/PowerShell/blob/master/.github/CONTRIBUTING.md#before-submitting)
        - [ ] [Add `[feature]` if the change is significant or affects feature tests](https://github.com/PowerShell/PowerShell/blob/master/docs/testing-guidelines/testing-guidelines.md#requesting-additional-tests-for-a-pr)

I'm unsure how to meaningfully add tests to prevent regression for this

IISResetMe committed:
Conditionally mark getter/setter implementations virtual

When implementing interfaces, PowerShell incorrectly produces non-virtual get/set methods for interface-defined properties.
This commit adds a lookup method for interface-defined properties and marks get/set methods for properties with matching signatures virtual.

IISResetMe commented:
@iSazonov if you have anything that's either more correct, complete or aesthetically pleasing, please feel free to update the PR :)

iSazonov commented:
@IISResetMe My post was informational. If you commit will approved I can grab it to my branch or rebase my branch.

IISResetMe committed:
Fix codefactor issues in ShouldImplementProperty

IISResetMe commented:
@iSazonov Fixed the CodeFactor issues in `ShouldImplementProperty()` but I'm unsure what to do with the "Complex Method" finding for `EmitPropertyIL()`

IISResetMe committed:
Add inheritance test for interface properties

iSazonov commented:
We ignore "Complex Method".

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1:
```diff
@@ -63,6 +63,14 @@ Describe 'Classes inheritance syntax' -Tags "CI" {
     }
 
+    It 'can implement .NET interface properties' {
+        Add-Type -TypeDefinition 'public interface I1 { int I{ get; set; } }'
```
Please replace "I" with more readable name. Maybe `InterfaceProperty`?

IISResetMe replied:
Update both type names and the property name to be more descriptive

iSazonov commented on test/powershell/Language/Classes/scripting.Classes.inheritance.tests.ps1:
```diff
@@ -63,6 +63,14 @@ Describe 'Classes inheritance syntax' -Tags "CI" {
+        Add-Type -TypeDefinition 'public interface I1 { int I{ get; set; } }'
+        $C1 = Invoke-Expression 'class C1 : I1 {[int]$I}'
+        $getter = $C1.GetType().GetMember('get_I')
+        $getter.ReturnType |Should -BeOfType System.Int32
```
Please add space before `Should`.

IISResetMe committed:
Rename interface in interface property test

iSazonov commented on src/System.Management.Automation/engine/parser/PSType.cs:
```diff
@@ -434,6 +434,19 @@ private Type GetBaseTypes(Parser parser, TypeDefinitionAst typeDefinitionAst, ou
 
+            private bool ShouldImplementProperty(string name, Type type)
+            {
+                foreach (var @interface in _typeBuilder.GetInterfaces())
```
I'd not use a reserved name. Maybe `interfaceType`?

IISResetMe committed:
Avoid reserved keyword in variable name interface

IISResetMe committed:
Emit test class during interface inheritance test

Build tests currently failing because the interface property test is missing
an instance of the test class to inspect

IISResetMe committed:
Fix type check in interface inheritance test

IISResetMe committed:
Fix interface inheritance test (again)

iSazonov commented:
Reopen to restart Appveyor CI.,

iSazonov commented:
@IISResetMe Thanks for your contribution!
